In [ ]:
import os
# Find the latest version of spark 3.x  from http://www.apache.org/dist/spark/ and enter as the spark version
# For example:
# spark_version = 'spark-3.4.0'
spark_version = 'spark-3.4.2'
os.environ['SPARK_VERSION']=spark_version

# Install Spark and Java
!apt-get update
!apt-get install openjdk-11-jdk-headless -qq > /dev/null
!wget -q http://www.apache.org/dist/spark/$SPARK_VERSION/$SPARK_VERSION-bin-hadoop3.tgz
!tar xf $SPARK_VERSION-bin-hadoop3.tgz
!pip install -q findspark

# Set Environment Variables
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-11-openjdk-amd64"
os.environ["SPARK_HOME"] = f"/content/{spark_version}-bin-hadoop3"

# Start a SparkSession
import findspark
findspark.init()

Hit:1 http://security.ubuntu.com/ubuntu jammy-security InRelease
Hit:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease
Hit:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Hit:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease
Hit:6 http://archive.ubuntu.com/ubuntu jammy-backports InRelease
Hit:7 https://ppa.launchpadcontent.net/c2d4u.team/c2d4u4.0+/ubuntu jammy InRelease
Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Reading package lists... Done


In [ ]:
# Import packages
from pyspark.sql import SparkSession
import time

# Create a SparkSession
spark = SparkSession.builder.appName("SparkSQL").getOrCreate()

In [ ]:
# 1. Read in the AWS S3 bucket into a DataFrame.
from pyspark import SparkFiles
url = "https://2u-data-curriculum-team.s3.amazonaws.com/dataviz-classroom/v1.2/22-big-data/home_sales_revised.csv"
spark.sparkContext.addFile(url)
df = spark.read.csv(SparkFiles.get("home_sales_revised.csv"), sep=",", header=True)

# Show DataFrame
df.show()

+--------------------+----------+----------+------+--------+---------+-----------+--------+------+----------+----+
|                  id|      date|date_built| price|bedrooms|bathrooms|sqft_living|sqft_lot|floors|waterfront|view|
+--------------------+----------+----------+------+--------+---------+-----------+--------+------+----------+----+
|f8a53099-ba1c-47d...|2022-04-08|      2016|936923|       4|        3|       3167|   11733|     2|         1|  76|
|7530a2d8-1ae3-451...|2021-06-13|      2013|379628|       2|        2|       2235|   14384|     1|         0|  23|
|43de979c-0bf0-4c9...|2019-04-12|      2014|417866|       2|        2|       2127|   10575|     2|         0|   0|
|b672c137-b88c-48b...|2019-10-16|      2016|239895|       2|        2|       1631|   11149|     2|         0|   0|
|e0726d4d-d595-407...|2022-01-08|      2017|424418|       3|        2|       2249|   13878|     2|         0|   4|
|5aa00529-0533-46b...|2019-01-30|      2017|218712|       2|        3|       196

In [ ]:
# 2. Create a temporary view of the DataFrame.
df.createOrReplaceTempView('v_sales')


In [ ]:
# 3. What is the average price for a four bedroom house sold in each year rounded to two decimal places?

##temporary view

SQL = """select
    year(date) as yearSold,
     round(avg(price),2)
    from v_sales
    where bedrooms = 4
    group by yearSold
    order by yearSold"""


spark.sql(SQL).show()

+--------+--------------------+
|yearSold|round(avg(price), 2)|
+--------+--------------------+
|    2019|            300263.7|
|    2020|           298353.78|
|    2021|           301819.44|
|    2022|           296363.88|
+--------+--------------------+



In [ ]:
# 4. What is the average price of a home for each year the home was built that have 3 bedrooms and 3 bathrooms rounded to two decimal places?

##temporary view

SQL = """select
    year(date) as yearSold,
    round(avg(price),2)
    from v_sales
    where bedrooms = 3
    and bathrooms = 3
    group by yearSold
    order by yearSold"""


spark.sql(SQL).show()


+--------+--------------------+
|yearSold|round(avg(price), 2)|
+--------+--------------------+
|    2019|           287287.82|
|    2020|           294204.16|
|    2021|           294211.46|
|    2022|           292725.69|
+--------+--------------------+



In [ ]:
# 5. What is the average price of a home for each year built that have 3 bedrooms, 3 bathrooms, with two floors,
# and are greater than or equal to 2,000 square feet rounded to two decimal places?

##temporary view

SQL = """select
    year(date) as yearSold,
    round(avg(price),2)
    from v_sales
    where bedrooms = 3
    and bathrooms = 3
    and floors = 2
    and sqft_living > 2000
    group by yearSold
    order by yearSold"""


spark.sql(SQL).show()

+--------+--------------------+
|yearSold|round(avg(price), 2)|
+--------+--------------------+
|    2019|           289976.57|
|    2020|           292289.09|
|    2021|           296330.96|
|    2022|           289517.82|
+--------+--------------------+



In [ ]:
# 6. What is the "view" rating for the average price of a home, rounded to
#two decimal places, where the homes are greater than
# or equal to $350,000? Although this is a small dataset, determine the run time for this query.


##temporary view timings

start_time = time.time()

SQL = """select
    year(date) as yearSold,
    sum(view) as viewRating,
    round(avg(price),2) as Price
    from v_sales
    where
    Price >= 350000
    group by yearSold
    order by yearSold"""


spark.sql(SQL).show()

print("--- %s seconds ---" % (time.time() - start_time))

+--------+----------+---------+
|yearSold|viewRating|    Price|
+--------+----------+---------+
|    2019|  104902.0|478628.66|
|    2020|  102832.0|476026.01|
|    2021|  105142.0|468070.46|
|    2022|   46061.0|471010.02|
+--------+----------+---------+

--- 0.942920446395874 seconds ---


In [ ]:
# 7. Cache the the temporary table home_sales.
spark.sql("cache table v_sales")

DataFrame[]

In [ ]:
# 8. Check if the table is cached.
spark.catalog.isCached('v_sales')

True

In [ ]:
# 9. Using the cached data, run the query that filters out the view ratings with average price
#  greater than or equal to $350,000. Determine the runtime and compare it to uncached runtime.

start_time = time.time()
##this is cached timing


SQL = """select
    year(date) as yearSold,
    sum(view) as viewRating,
    round(avg(price),2) as Price
    from v_sales
    where
    Price >= 350000
    group by yearSold
    order by yearSold"""

print("--- %s seconds ---" % (time.time() - start_time))


--- 0.00011706352233886719 seconds ---


Determine the runtime and compare it to uncached runtime.


*   THE RUNTIME FOR THE UNCACHED IS --- 0.942920446395874 seconds ---
*   THE RUNTIME FOR THE CACHED IS ---   0.00011706352233886719 seconds -

CONCLUSION: CACHED IS FASTER THEN UNCACHED


In [ ]:
# 10. Partition by the "date_built" field on the formatted parquet home sales data
df.write.partitionBy("date_built").mode("overwrite").parquet("date_built_partitioned")

In [ ]:
# 11. Read the parquet formatted data.
p_df_dtbuiltp=spark.read.parquet('date_built_partitioned')

In [ ]:
# 12. Create a temporary table for the parquet data.
p_df_dtbuiltp.createOrReplaceTempView('p_dtbuilt_tb_p')

In [ ]:
# 13. Run the query that filters out the view ratings with average price of greater than or equal to $350,000
# with the parquet DataFrame. Round your average to two decimal places.
# Determine the runtime and compare it to the cached version.


start_time = time.time()

SQL = """select
    year(date) as yearSold,
    sum(view) as viewRating,
    round(avg(price),2) as Price
    from p_dtbuilt_tb_p
    where
    Price >= 350000
    group by yearSold
    order by yearSold"""

print("--- %s seconds ---" % (time.time() - start_time))

--- 0.00013208389282226562 seconds ---


Determine the runtime and compare it to uncached runtime.

THE RUNTIME FOR THE UNCACHED IS --- 0.942920446395874 seconds ---
THE RUNTIME FOR THE CACHED IS   ---0.00013208389282226562 seconds --

CONCLUSION: CACHED IS FASTER THEN UNCACHED

In [ ]:
# 14. Uncache the home_sales temporary table.
spark.sql("uncache table v_sales")

DataFrame[]

In [ ]:
# 15. Check if the home_sales is no longer cached

spark.catalog.isCached("v_sales")

False